src/Package/Impl/ProjectSystem/Commands/SendFileCommandBase.cs

```diff
@@ -32,70 +33,81 @@ internal class SendFileCommandBase {
+                    }), ct);
                 });
 
-                using (var fts = new DataTransferSession(session, _fs)) {
-                    cookie = 0;
-                    statusBar.SetText(Resources.Info_TransferringFiles);
+                if (ct.IsCancellationRequested) {
+                    return;
+                }
 
-                    total = (uint)_fs.FileSize(compressedFilePath);
+                using (var fts = new DataTransferSession(session, _fs)) {
+                    long size = _fs.FileSize(compressedFilePath);
 
+                    // Transfer phase: 2 of 3 phases
                     var remoteFile = await fts.SendFileAsync(compressedFilePath, true, new Progress<long>((b) => {
-                        statusBar.Progress(ref cookie, 1, Resources.Info_TransferringFiles, (uint)b, total);
-                    }), cancellationToken);
+                        int step = 33; // start with 33% to indicate compression phase is done.
+                        step += (int)((((double)b / (double)size) * 100) / 3); // divide by 3, this is for transfer phase.
+                        progress.Report(new ProgressDialogData(step, Resources.Info_TransferringFilesWithSize.FormatInvariant(b, size)));
+                    }), ct);
+
+                    if (ct.IsCancellationRequested) {
+                        return;
+                    }
 
-                    statusBar.SetText(Resources.Info_ExtractingFilesInRHost);
-                    await session.EvaluateAsync<string>($"rtvs:::save_to_project_folder({remoteFile.Id}, {projectName.ToRStringLiteral()}, '{remotePath.ToRPath()}')", REvaluationKind.Normal, cancellationToken);
+                    // Extract phase: 3 of 3 phases
+                    // start with 66% completion to indicate compression and transfer phases are done.
+                    progress.Report(new ProgressDialogData(66, Resources.Info_ExtractingFilesInRHost));
+                    await session.EvaluateAsync<string>($"rtvs:::save_to_project_folder({remoteFile.Id}, {projectName.ToRStringLiteral()}, '{remotePath.ToRPath()}')", REvaluationKind.Normal, ct);
 
-                    _appShell.DispatchOnUIThread(() => {
-                        outputWindow.WriteLine(Resources.Info_TransferringFilesDone);
-                    });
+                    progress.Report(new ProgressDialogData(100, Resources.Info_TransferringFilesDone));
+
+                    paths.ForEach((s) => _appShell.DispatchOnUIThread(() => {
+                        outputWindow.WriteLine(s);
+                    }));
                 }
-            } catch(Exception ex) when (ex is UnauthorizedAccessException || ex is IOException) {
-                _appShell.ShowErrorMessage(Resources.Error_CannotTransferFile.FormatInvariant(ex.Message));
+            } catch (TaskCanceledException) {
+                _appShell.DispatchOnUIThread(() => {
```

Comment:
If you use `new InteractiveWindowConsole()` and use `IConsole` you won't have to worry about threading.